[ENH] Implemented Echo State Network (ESN) Classifier

sharma-kshitij-ks opened the pull request:
This commit implements the Echo State Network (ESN) classifier in the deep_learning category of sktime. 
Currently I realised that the sktime codebase lacked an implementation for the ESN classifier, which is a type of reservoir computing model and a deep learning architecture.

The format mentioned in extension templates has been followed and it's imperative that this will be a good starting point for further developement.


#1147 Under The Category [Pick an algorithm to implement or interface, according to the [extension templates](https://github.com/sktime/sktime/tree/main/extension_templates). Possible choices are forecasters (univariate or multivariate), time series classifiers, time series regressors, annotators, distances, kernels - these can be existing feature requests, or new suggestions. We currently do not recommend transformers as they undergo refactoring.]

Thank You for reviewing, please open an Issue to discuss changes for the same.

fkiraly commented on sktime/classification/deep_learning/esn.py:
```diff
@@ -0,0 +1,155 @@
+    _tags = {
+        "authors": ["sharma-kshitij-ks"],
+        "maintainers": ["sharma-kshitij-ks"],
+        "python_dependencies": "numpy, scipy, scikit-learn",
```
these are core deps, not necessary to add

fkiraly reviewed (changes requested):
Nice! I was not familiar with ESN, but it looks like this indeed faithfully implements the algorithm in the mentioned paper.

Some change requests:

* For implementing, kindly ensure that you follow the extension template fore classifcation (in `extension_templates`) - e.g., implementing `_fit` and not `fit`
* the paper makes some recommendations for sequentes or time series classifcation, in section 4.7 - should these be an option at least?
* I think it should be possible to pick any classifier that is sklearn compatible. If none is chosend (default, None), it should default to RidgeClassifier, as in the paper.

fkiraly commented:
Also pinging @frthjf as he was interested in reservoir computing a while ago.

frthjf commented:
Thanks! Not a code review, but I think this is a great addition. I was working on [another reservoir computing model](https://github.com/sktime/sktime/issues/4521) a while back; when I find some time, I'll see if I can finish a PR for that as well.

sharma-kshitij-ks commented:
Here's a summary of the changes made in the commit:

Implemented the _fit and _predict methods instead of fit and predict to align with the sktime extension template for classification.
Added the option to choose a base classifier via the base_classifier parameter, defaulting to RidgeClassifier if no classifier is specified. This aligns with the paper's recommendation and offers flexibility to use any sklearn-compatible classifier.

sharma-kshitij-ks commented:
Here's the summary of the changes made in the commit: "Updated esn.py" 

Implemented the _fit and _predict methods instead of fit and predict to align with the sktime extension template for classification.
Added the option to choose a base classifier via the base_classifier parameter, defaulting to RidgeClassifier if no classifier is specified. This aligns with the paper's recommendation and offers flexibility to use any sklearn-compatible classifier.

fkiraly commented on sktime/classification/deep_learning/esn.py:
```diff
@@ -0,0 +1,159 @@
+    _tags = {
+        "authors": ["sharma-kshitij-ks"],
+        "maintainers": ["sharma-kshitij-ks"],
+        "python_dependencies": "core",
```
if there are no additional deps, then you do not need to set this tag.

fkiraly commented on sktime/classification/deep_learning/esn.py:
```diff
@@ -0,0 +1,159 @@
+            creates a valid test instance.
+            ``create_test_instance`` uses the first (or only) dictionary in ``params``.
+        """
+        params = {"n_reservoir": 50, "regularization": 1e-4}
```
please ensure you have at least two, better three parameter sets, covering diffeerent settings of your params. The empty dict (all defaults) should be included

fkiraly commented on sktime/classification/deep_learning/esn.py:
```diff
@@ -0,0 +1,159 @@
+        leaking_rate=0.3,
+        regularization=1e-6,
+        random_state=None,
+        base_classifier=None,
```
base_classifier should be the first param, since it is a component

fkiraly commented on sktime/classification/deep_learning/esn.py:
```diff
@@ -0,0 +1,159 @@
+        Regularization parameter for ridge regression.
+    random_state: int or None, default=None
+        Seed for random number generator.
+    base_classifier: sklearn classifier, default=None
```
default should be stated here - i.e., the internal default (whic hclassifier)

fkiraly commented on sktime/classification/deep_learning/esn.py:
```diff
@@ -0,0 +1,159 @@
+class EchoStateNetwork(BaseClassifier):
+    """Echo State Network (ESN) Classifier.
+
+    Overview:
```
this should refer to the key reference, optimally how parameters of this implementation map to those in the paper.

You can number your references by using `[1]` etc in restructured text.

fkiraly commented on sktime/classification/deep_learning/esn.py:
```diff
@@ -0,0 +1,159 @@
+__author__ = ["sharma-kshitij-ks"]
+
+from sklearn.base import BaseEstimator, ClassifierMixin
```
please ensure you adhere to code formatting standards, the guide is here:
https://www.sktime.net/en/stable/developer_guide/coding_standards.html

fkiraly reviewed (changes requested):
Great!

I've left some comments above.

Here is a more extensive guide on adding estimators:
https://www.sktime.net/en/latest/developer_guide/add_estimators.html

Question, have you tried to use `check_estimator`? It does not look like your estimator would currently pass. The link explains how to use this for testing.

sharma-kshitij-ks commented:
Still needs revision, especially check_estimator compatibility & code formatting standard request(s) still pending, will get back to it.
All other suggestions incorporated.

sharma-kshitij-ks commented:
Still needs revision, especially testing check_estimator compatibility, adding paper references/structure & code formatting standard request(s) are still pending, will get back to it.
All other suggestions incorporated.
Did some modifications in _fit aswell - about estimator initialisation. Please review again.

fkiraly commented:
thanks. Waiting for the `check_estimator` tests passing, will review then. Please let us know if you need advice for that.

fkiraly commented:
@sharma-kshitij-ks, from the state of your code I would assume you have not actually run it. I would advise to start with `check_estimator` testing for conformance.
